File: lint-rules-android-lint/src/main/java/com/vanniktech/lintrules/android/WrongViewIdFormatDetector.java
package com.vanniktech.lintrules.android;

import com.android.annotations.NonNull;
import com.android.tools.lint.detector.api.Implementation;
import com.android.tools.lint.detector.api.Issue;
import com.android.tools.lint.detector.api.LayoutDetector;
import com.android.tools.lint.detector.api.XmlContext;
import java.util.Collection;
import org.w3c.dom.Attr;

import static com.android.SdkConstants.ATTR_ID;
import static com.android.tools.lint.detector.api.Category.CORRECTNESS;
import static com.android.tools.lint.detector.api.Scope.RESOURCE_FILE_SCOPE;
import static com.android.tools.lint.detector.api.Severity.WARNING;
import static java.util.Collections.singletonList;

public final class WrongViewIdFormatDetector extends LayoutDetector {
  static final Issue ISSUE_WRONG_VIEW_ID_FORMAT = Issue.create("WrongViewIdFormat", "Ids should be in lowerCamelCase Format.",
      "Ids should be in lowerCamelCase Format.", CORRECTNESS, 8, WARNING,
      new Implementation(WrongViewIdFormatDetector.class, RESOURCE_FILE_SCOPE));

  @Override public Collection<String> getApplicableAttributes() {
    return singletonList(ATTR_ID);
  }

  @Override public void visitAttribute(@NonNull final XmlContext context, @NonNull final Attr attribute) {
    if (ATTR_ID.equals(attribute.getLocalName())) {
      final String id = attribute.getValue().replace("@+id/", "");
      final boolean isSuppressed = context.getDriver().isSuppressed(context, ISSUE_WRONG_VIEW_ID_FORMAT, attribute);

      if (!isCamelCase(id) && !isSuppressed) {
        context.report(ISSUE_WRONG_VIEW_ID_FORMAT, context.getValueLocation(attribute), "Id is not in lowerCamelCaseFormat");
      }
    }
  }

  static boolean isCamelCase(final String string) {
    return !Character.isUpperCase(string.charAt(0)) && !string.contains("_");
  }
}

File: lint-rules-android-lint/src/main/java/com/vanniktech/lintrules/android/WrongViewIdFormatDetector.kt
package com.vanniktech.lintrules.android

import com.android.SdkConstants.ATTR_ID
import com.android.tools.lint.detector.api.Category.Companion.CORRECTNESS
import com.android.tools.lint.detector.api.Implementation
import com.android.tools.lint.detector.api.Issue
import com.android.tools.lint.detector.api.LayoutDetector
import com.android.tools.lint.detector.api.LintUtils
import com.android.tools.lint.detector.api.Scope.Companion.RESOURCE_FILE_SCOPE
import com.android.tools.lint.detector.api.Severity.WARNING
import com.android.tools.lint.detector.api.XmlContext
import org.w3c.dom.Attr

val ISSUE_WRONG_VIEW_ID_FORMAT = Issue.create("WrongViewIdFormat",
    "Flag view ids that are not in lowerCamelCase Format.",
    "View ids should be in lowerCamelCase format. This has the benefit of saving an unnecessary underscore and also just looks nicer.",
    CORRECTNESS, PRIORITY, WARNING,
    Implementation(WrongViewIdFormatDetector::class.java, RESOURCE_FILE_SCOPE))

class WrongViewIdFormatDetector : LayoutDetector() {
  override fun getApplicableAttributes() = listOf(ATTR_ID)

  override fun visitAttribute(context: XmlContext, attribute: Attr) {
    if (!LintUtils.stripIdPrefix(attribute.value).isLowerCamelCase()) {
      val fix = fix().replace()
          .name("Convert to lowerCamelCase")
          .text(attribute.value)
          .with(attribute.value.idToSnakeCase())
          .build()

      context.report(ISSUE_WRONG_VIEW_ID_FORMAT, attribute, context.getValueLocation(attribute), "Id is not in lowerCamelCaseFormat", fix)
    }
  }
}
